### ISSUE TITLE
Deserialization issue with AssetArchive / StringAsset

### ISSUE THREAD

**Issue by dferretti** [pulumi/pulumi#7076]
Using the C# SDK, I am trying to create an AWS Lambda function from an `Output<string>` created earlier in the stack representing the javascript source code of the function. In one particular stack I am working on, I am consistently seeing serialization errors on subsequent `up` or `preview` operations after an initial successful `up`. 

I have tried to whittle my stack down to just the resources in play here, but this example stack actually is able to deploy without any issue 🤔 so I am leaving it here until I can come back with a reproducible example, in hopes this looks familiar to somebody in the meantime. 

## Steps to reproduce
Here is a representative stack example - again, this stack has no issue but is created pretty much identically to my problematic resource inside my larger stack. I will replace this example with a better one once I can nail down what causes it to fail.
```c#
var role = new Pulumi.Aws.Iam.Role("role", new Pulumi.Aws.Iam.RoleArgs
{
    Name = "pulumi-issue-role",
    AssumeRolePolicy = @"{""Version"":""2008-10-17"",""Statement"":[{""Effect"":""Allow"",""Principal"":{""Service"":[""lambda.amazonaws.com"",""edgelambda.amazonaws.com""]},""Action"":""sts:AssumeRole""}]}",
});

Output<string> handlerCode = ... // based on outputs defined earlier in the stack

var handlerArchive = handlerCode.Apply(code => (Archive)new AssetArchive(new Dictionary<string, AssetOrArchive>
{
    ["index.js"] = new StringAsset(code),
}));

var function = new Pulumi.Aws.Lambda.Function("function", new Pulumi.Aws.Lambda.FunctionArgs
{
    Code = handlerArchive,
    Handler = "index.handler",
    Role = role.Arn,
    Runtime = Pulumi.Aws.Lambda.Runtime.NodeJS14dX,
    Publish = true,
});
```
When I first `pulumi up` this stack it works fine. The next time I try it, I get this serialization error:
```
System.InvalidOperationException: Expected Pulumi.Archive but got System.String deserializing Pulumi.Aws.Lambda.Function.code
   at Pulumi.Serialization.Converter.ConvertObject(String context, Object val, Type targetType)
   at Pulumi.Serialization.Converter.ConvertValue(String context, Value value, Type targetType, ImmutableHashSet`1 resources)
   at Pulumi.Deployment.CompleteResourceAsync(Resource resource, Boolean remote, Func`2 newDependency, ResourceArgs args, ResourceOptions options, ImmutableDictionary`2 completionSources)
   at Pulumi.Deployment.Runner.<>c__DisplayClass9_0.<<WhileRunningAsync>g__HandleCompletion|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Pulumi.Deployment.Runner.WhileRunningAsync()<{%reset%}>)
```
On this example stack and others that work fine, I have noticed in the state file the `code` input/output of the lambda function is serialized like this: (identical in the `inputs` and `outputs` section of the state for the lambda function)
```json
"type": "aws:lambda/function:Function",
"inputs": {
    ...
    "code": {
        "4dabf18193072939515e22adb298388d": "0def7320c3a5731c473e5ecbe6d01bc7",
        "assets": {
            "index.js": {
                "4dabf18193072939515e22adb298388d": "c44067f5952c0a294b673a41bacd8c17",
                "hash": "5094...",
                "text": "..."
            }
        },
        "hash": "049f885f54552ba53ad1a1c8c477c3681be05baf8e6fb7465fc5a2fdf72ffe18"
    },
}
```
However, for the problematic stack, the `inputs.code` section looks the same as above, but the `outputs` is serialized like this:
(the file name matches the hash of the input, if that is relevant)
```json
"code": "/tmp/pulumi-asset-049f885f54552ba53ad1a1c8c477c3681be05baf8e6fb7465fc5a2fdf72ffe18",
```

Using the `pulumi/pulumi:v3.2.1` docker container and `Pulumi.Aws v4.3.0` nuget package.

**Comment by dferretti** [comment]
Ok I finally have something reproducible - it is actually the `refresh` operation that seems to be the culprit. The `Output<string>` is not relevant. Here is a more simplified complete example:
```c#
public class Program
{
    public static async Task Main(string[] args)
    {
        await Deployment.RunAsync<LambdaStack>();
    }
}

public class LambdaStack : Stack
{
    public LambdaStack()
    {
        var role = new Pulumi.Aws.Iam.Role("role", new Pulumi.Aws.Iam.RoleArgs
        {
            Name = "pulumi-issue-role",
            AssumeRolePolicy = @"{""Version"":""2008-10-17"",""Statement"":[{""Effect"":""Allow"",""Principal"":{""Service"":[""lambda.amazonaws.com"",""edgelambda.amazonaws.com""]},""Action"":""sts:AssumeRole""}]}",
        });

        var handlerArchive = new AssetArchive(new Dictionary<string, AssetOrArchive>
        {
            ["index.js"] = new StringAsset("console.log('hello world');"),
        });

        var function = new Pulumi.Aws.Lambda.Function("function", new Pulumi.Aws.Lambda.FunctionArgs
        {
            Code = handlerArchive,
            Handler = "index.handler",
            Role = role.Arn,
            Runtime = Pulumi.Aws.Lambda.Runtime.NodeJS14dX,
        });
    }
}
```

After a first successful `pulumi up`, the state file looks like this: https://gist.github.com/dferretti/4ec9eff9df10705e71073afb6e87daa0#file-state-1-json
The important part is that both `inputs` and `outputs` contain this:
```json
"code": {
	"4dabf18193072939515e22adb298388d": "0def7320c3a5731c473e5ecbe6d01bc7",
	"assets": {
		"index.js": {
			"4dabf18193072939515e22adb298388d": "c44067f5952c0a294b673a41bacd8c17",
			"hash": "fdbfc7bd27940b250bd17955d592b49f00e4d7c3b606933949c8fe759d707886",
			"text": "console.log('hello world');"
		}
	},
	"hash": "e0a492485864fa2a43dc12085a2e8f81c9470faedfb2d676c0c14b4aa987718c"
},
```
I can then `pulumi up` and `pulumi preview` repeatedly without issue. However, once I run `pulumi refresh`, it detects a change in the function code:
```
Refreshing (test):
     Type                    Name       Status      Info
     pulumi:pulumi:Stack     test-test
     ├─ aws:iam:Role         role
 ~   └─ aws:lambda:Function  function   updated     [diff: ~code]

Resources:
    ~ 1 updated
    2 unchanged

Duration: 1s
```
and after it is done, the state file looks like: https://gist.github.com/dferretti/4ec9eff9df10705e71073afb6e87daa0#file-state-2-json
And the important part is that both the `inputs`  and `outputs` look like
```json
"code": "/tmp/pulumi-asset-e0a492485864fa2a43dc12085a2e8f81c9470faedfb2d676c0c14b4aa987718c",
```

After this point, any `pulumi preview` or `pulumi up` will fail with:
```
root@d3e40046faf9:/app# pulumi preview
Previewing update (test):
     Type                 Name       Plan     Info
     pulumi:pulumi:Stack  test-test           1 error

Diagnostics:
  pulumi:pulumi:Stack (test-test):
    error: Running program '/app/bin/Debug/netcoreapp3.1/PulumiIssue.dll' failed with an unhandled exception:
    System.InvalidOperationException: Expected Pulumi.Archive but got System.String deserializing Pulumi.Aws.Lambda.Function.code
       at Pulumi.Serialization.Converter.ConvertObject(String context, Object val, Type targetType) in /_/sdk/dotnet/Pulumi/Serialization/Converter.cs:line 44
       at Pulumi.Serialization.Converter.ConvertValue(String context, Value value, Type targetType, ImmutableHashSet`1 resources) in /_/sdk/dotnet/Pulumi/Serialization/Converter.cs:line 33
       at Pulumi.Deployment.CompleteResourceAsync(Resource resource, Boolean remote, Func`2 newDependency, ResourceArgs args, ResourceOptions options, ImmutableDictionary`2 completionSources) in /_/sdk/dotnet/Pulumi/Deployment/Deployment_ReadOrRegisterResource.cs:line 120
       at Pulumi.Deployment.Runner.<>c__DisplayClass9_0.<<WhileRunningAsync>g__HandleCompletion|0>d.MoveNext() in /_/sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs:line 131
    --- End of stack trace from previous location where exception was thrown ---
       at Pulumi.Deployment.Runner.WhileRunningAsync() in /_/sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs:line 171
```

**Comment by dferretti** [comment]
It looks somewhat similar to https://github.com/pulumi/pulumi/issues/6691 but not exactly the same

**Comment by dferretti** [comment]
Ah - this looks relevant: https://github.com/pulumi/pulumi-aws/blob/v4.3.0/provider/resources.go#L1741-L1764

I am definitely outside my comfort zone here, but the comments there and the tfbridge code it points to here https://github.com/pulumi/pulumi-terraform-bridge/blob/master/pkg/tfbridge/assets.go#L167-L171

match up with the behavior I'm seeing - where my in-memory archive is getting replaced with a temp zip. I'm kinda guessing here but the problem seems to be that it gets serialized in the state file as just a naked string instead of a FileAsset with a json object containing sig/hash/path etc. Maybe the tfbridge is supposed to take care of that though, I don't know where it comes into play.

**Comment by dferretti** [comment]
After seeing https://github.com/pulumi/pulumi/pull/7579 and https://github.com/pulumi/pulumi-aws/issues/1521#issuecomment-887146069 and bumping both Pulumi CLI to 3.12.0 and Pulumi.Aws to 4.21.0 it looks like everything is working as it should. I am able to `refresh` correctly now :+1:

**Comment by zyv** [comment]
So for us the problem is unfortunately not fixed. If we do `refresh` after up, we see that `refresh` is trying to corrupt the state file:

```
    ~ aws:lambda/function:Function: (update)
        [id=move-cloudfront-logs-for-athena-global]
        [urn=urn:pulumi:global::world::aws:lambda/function:Function::lambda-move-cloudfront-logs-for-athena-global]
        [provider=urn:pulumi:global::world::pulumi:providers:aws::default_4_38_1::d33be2bf-0232-4a23-84d9-e9b3e895bf60]
        --outputs:--
      - code                        : archive(assets:d5d8371) {
      -     ".": archive(file:d5d8371) { lambda_functions/global_stack/move_cloudfront_logs_for_athena }
        }
      + code                        : "/tmp/pulumi-asset-d5d837160b9e4ab4811d0ee9e2a6e2c4b8397fc91acf64c4801679e8109be55d"
```

We create the lambda using `FileArchive`.

```
$ pulumi version
v3.35.3
```

The version of `pulumi-aws` is `4.38.1`.

/cc @mm-matthias @marns93

**Comment by zyv** [comment]
@lukehoban @pgavlin should your fixes from #7579 somehow be ported to Python SDK or it was supposed to happen automagically and some more fixing is needed?
